### PR TITLE
combineObject()

### PR DESCRIPTION
**Pull request by uniyalprashant9**
We can call combineObject(path, keys_to_be_removed) as : `combineObject('/abc/pqr/',['id',img,])`

This will read all files in the given `path` and remove the keys given the list of ` keys_to_be_removed`.

#567

### CODE REVIEW

**Comment by atherdon** [comment]
this way will help only if we want to remove some fields(keys) from an
object.
but what if we'll completely change the logic of those objects.

i still think that using callback is a better way to handle this kind
of problem
your approach will work on some of the cases, but not of all of them.

check this
https://codeburst.io/javascript-what-the-heck-is-a-callback-aba4da2deced

**Comment by uniyalprashant9** [comment]
so we need to convert our json files to other json files that have a diffrent schema??

**Comment by atherdon** [comment]
@vadim9999 tell me what do you think about this changes

**Comment by uniyalprashant9** [comment]
@atherdon  should I do a readme update...for combineObject if you are ok with current changes and for splitObjcect also, as someone changed readme and inserted his update info in between my changes info it got mixed up.  [pull_request-545](https://github.com/GroceriStar/sd/pull/546)

yellow marked are mine changes and the red one are added by red are done by @Edebo  and seems incomplete as the last sentence is not complete.

Annotation 2019-06-25 181115




I searched many pages on google but didn't find any way to change json structure on a given structure in runtime however I can make a set of different fixed structures and the user can choose that structure if he needed.

**Comment by atherdon** [comment]
yes, please update a readme.

**Comment by uniyalprashant9** [comment]
> Code Climate has analyzed commit [a3154c9](https://github.com/GroceriStar/sd/commit/a3154c9377fc8ba813e91874b8eba4a702650d0f) and detected **1 issue** on this pull request.
> 
> Here's the issue category breakdown:
> 
> Category	Count
> Duplication	1
> View more on [Code Climate](https://codeclimate.com/github/GroceriStar/sd/pull/577).

I think this is falsely detected, I have checked both the files. It detected export blocks that have different modules in 2 files, the only similar thing is that both export modules have is the same number of export modules.

**Comment by uniyalprashant9** [comment]
image

@atherdon  What happed to these files. I didn't made any changes in them.

**Comment by atherdon** [comment]
maybe you can sync your fork? and it will show that changes. because as we have few members, working on the same files - changes came very often

**Comment by vadim9999** [comment]
After executing this code
```
const src_utils = require('../../src/utils')
const path = "/home/vadim/work/javascript/sd/src/data/Dash/"
var content = src_utils.readAllFiles(path)
console.log(content);
```
I get error
```
undefined:1
#### Dash
^

SyntaxError: Unexpected token # in JSON at position 0
```
1. Maybe I set doesn't the right path to a folder. For example I want to set path on folder "/home/vadim/work/javascript/sd/src/data/Dash/". Folder Dash has one file dash.json. 
2. I don't find information or an example of the correct path that should be set.
3. Is it tested someone this method readAllFiles()?
4. Is it a bug?

**Comment by vadim9999** [comment]
After deeper studying this method
 ```
let data = fs.readFileSync(path + file)
```
in variable data we have 
```
<Buffer 23 23 23 23 20 44 61 73 68 0a 0a 45 78 61 6d 70 6c 65 0a 60 60 60 0a 22 6e 61 6d 65 22 3a 22 64 61 73 68 22 2c 0a 22 61 6d 65 72 69 63 61 6e 53 74 61 ... >
```
But error is in this line of code
```
data = JSON.parse(data)
```
In result JSON.parse expect to get string in argument but get class Buffer

**Comment by vadim9999** [comment]
have found what is the problem. In the same folder was Readme.md that is not json. I think we should add there checker of file type. It allows not to get the error that I had before in my two comments upper

**Comment by vadim9999** [comment]
```
const src_utils = require('../../src/utils')
const path = "/home/vadim/work/javascript/sd/src/data/Dash/"

function updateContent(content, keys) {
  var len = content.length
  for (var i = 0; i < len; i++) {
      for (var j = 0; j < keys.length; j++) {
          console.log(content[i][keys[j]]); // returns undefined
          
          delete content[i][keys[j]] 
      }
  }
  return content
}
const keys =['name', 'volume']
var content = src_utils.readAllFiles(path)
console.log(content);
content1 = updateContent(content, keys)
console.log(content1);
```
Here is result
```
[ [ { name: 'dash',
      americanStandart: '1/8 teaspoon',
      americanStandartOunces: null,
      volume: null,
      weight: null } ] ]
undefined
undefined
[ [ { name: 'dash',
      americanStandart: '1/8 teaspoon',
      americanStandartOunces: null,
      volume: null,
      weight: null } ] ]
```
As you see content[i][keys[j]]  it returns undefined. But when pass array like this it's working wine
```
content1 = updateContent([{
   name: 'dash',
      americanStandart: '1/8 teaspoon',
      americanStandartOunces: null,
      volume: null,
      weight: null 
}], keys)
```
All json`s files contains array. function readAllFiles() return array that contain content of json file. 
```
[
// contant of json
[
{
name:"Asparagys"}
]

]
```
Please fix updateContent() - it's doesn't work

**Comment by atherdon** [comment]
> https://github.com/GroceriStar/sd/pull/577#issuecomment-505491996

should be fixed by now

**Comment by vadim9999** [comment]
@uniyalprashant9 Am I right understood this method combineObject() finds all json in a folder then delete some properties that were set by the user and write into a single file?

**Comment by vadim9999** [comment]
Add into combineObject thirld arguments - callback and call it correctly in the method

**Comment by uniyalprashant9** [comment]
> @uniyalprashant9 Am I right understood this method combineObject() finds all json in a folder then delete some properties that were set by the user and write into a single file?

@vadim9999  yes, it does exactly the same you said :)

**Comment by uniyalprashant9** [comment]
> After executing this code
> 
> ```
> const src_utils = require('../../src/utils')
> const path = "/home/vadim/work/javascript/sd/src/data/Dash/"
> var content = src_utils.readAllFiles(path)
> console.log(content);
> ```
> 
> I get error
> 
> ```
> undefined:1
> #### Dash
> ^
> 
> SyntaxError: Unexpected token # in JSON at position 0
> ```
> 
> 1. Maybe I set doesn't the right path to a folder. For example I want to set path on folder "/home/vadim/work/javascript/sd/src/data/Dash/". Folder Dash has one file dash.json.
> 2. I don't find information or an example of the correct path that should be set.
> 3. Is it tested someone this method readAllFiles()?
> 4. Is it a bug?

don't make path constant.

**Comment by uniyalprashant9** [comment]
> I have found what is the problem. In the same folder was Readme.md that is not json. I think we should add there checker of file type. It allows not to get the error that I had before in my two comments upper

yes i got it, will add a checker

**Comment by uniyalprashant9** [comment]
> ```
> const src_utils = require('../../src/utils')
> const path = "/home/vadim/work/javascript/sd/src/data/Dash/"
> 
> function updateContent(content, keys) {
>   var len = content.length
>   for (var i = 0; i < len; i++) {
>       for (var j = 0; j < keys.length; j++) {
>           console.log(content[i][keys[j]]); // returns undefined
>           
>           delete content[i][keys[j]] 
>       }
>   }
>   return content
> }
> const keys =['name', 'volume']
> var content = src_utils.readAllFiles(path)
> console.log(content);
> content1 = updateContent(content, keys)
> console.log(content1);
> ```
> 
> Here is result
> 
> ```
> [ [ { name: 'dash',
>       americanStandart: '1/8 teaspoon',
>       americanStandartOunces: null,
>       volume: null,
>       weight: null } ] ]
> undefined
> undefined
> [ [ { name: 'dash',
>       americanStandart: '1/8 teaspoon',
>       americanStandartOunces: null,
>       volume: null,
>       weight: null } ] ]
> ```
> 
> As you see content[i][keys[j]] it returns undefined. But when pass array like this it's working wine
> 
> ```
> content1 = updateContent([{
>    name: 'dash',
>       americanStandart: '1/8 teaspoon',
>       americanStandartOunces: null,
>       volume: null,
>       weight: null 
> }], keys)
> ```
> 
> All json`s files contains array. function readAllFiles() return array that contain content of json file.
> 
> ```
> [
> // content of json
> [
> {
> name:"Asparagys"}
> ]
> 
> ]
> ```
> 
> Please fix updateContent() - it's doesn't work

New code :-
```
const src_utils = require('../../src/utils')
var path = "C:/Users/Prashant/Desktop/"

function updateContent(content, keys) {
    var len = content.length
    console.log(len)
    for (itr = 0; itr < len; itr++) {
        var elementLen = content[itr].length
        for (var i = 0; i < elementLen; i++) {
            for (var j = 0; j < keys.length; j++) {
                console.log(content[itr][i][keys[j]]);
                delete content[itr][i][keys[j]]
            }
        }
    }
    return content
}
const keys = ['id', 'desc']
var content = src_utils.readAllFiles(path)
    //console.log(content);
content1 = updateContent(content, keys)
console.log(content1);
```

result:-
```
1
25
false
26
false
[ [ { departments: [Array],
      name:
       '14-grocery-staples-I-always-have-as-a-20-something-on-a-budget',
      img: false,
      slug: false },
    { departments: [Array],
      name: 'building-a-healthy-vegan-grocery-list',
      img: false,
      slug: false } ] ]
```

**Comment by atherdon** [comment]
it will be awesome if all of this examples, that you put here will be
stored at readme for a future work at this package

>

**Comment by uniyalprashant9** [comment]
I think there is a problem with writeFile() module... it writes files but when I read those new files from readAllFiles () it gives:

image


but when I tried to read grocery.json in src/data/grocery it worked 

image

**Comment by uniyalprashant9** [comment]
when the content was this no white spaces and I tried to read this file from readAllFiles it gave an error:

image

but when I read this file which has white spaces in between it worked.

image

**Comment by uniyalprashant9** [comment]
I need to work on writeFile(), it must add whitespaces in json files or else it will give an error when we try to read new files.

**Comment by uniyalprashant9** [comment]
@atherdon  a few files are changed by someone, build fails. :(
I have updated fork already.
image

**Comment by uniyalprashant9** [comment]
> Code Climate has analyzed commit [ba02dbd](https://github.com/GroceriStar/sd/commit/ba02dbd4c40b1c3fecb5a5a163215e8fc0003248) and detected **2 issues** on this pull request.
> 
> Here's the issue category breakdown:
> 
> Category	Count
> Complexity	1
> Duplication	1
> View more on [Code Climate](https://codeclimate.com/github/GroceriStar/sd/pull/577).

will fix the complexity issue when all modules are in the wroking state :)

**Comment by atherdon** [comment]
> I need to work on writeFile(), it must add whitespaces in json files or else it will give an error when we try to read new files.

sure please do it, improve writefile, but remember that as much shorter it can be - it's better



> @atherdon a few files are changed by someone, build fails. :(
> I have updated fork already.

Sure, sometimes it happens, it's a sign that we should handle this PR quickly and merge it - because i think that last tasks giving you hard times :) which is not bad, but they don't was so easy as it looks on the first views. But i like your progress - keep going!

**Comment by atherdon** [comment]
please sync your fork with master, i think i fix that error https://github.com/GroceriStar/sd/pull/577#issuecomment-506111702

**Comment by uniyalprashant9** [comment]
If I read grocery.json and split it from the modules I made it works fine.
image

the 2 files generated are:
image
image

and when I try to read these 2 new files it also works fine 
image

BUT WHEN I TRY TO SPLIT `grocery.json` AND TRY TO READ THE SPLITTED FILES IN ONE GO THE ERROR COMES and the content of splitted files disappears :(

image

I didn't understand why this is happening so, please help :(
I have pushed the changes in this PR.
@atherdon @vadim9999

File: generator/writeFile.js [hunk 1]
```diff
@@ -75,42 +74,51 @@ function readData (path, file) {
...
-function splitObject (path, file, flag = 1) { // split large files into single elements
-  /*
-      flag=1 ==> name according to index
-      flag=0 ==> name according to "name" attribute
-    */
-  path = fixPath(path)
-  let fileData = readData(path, file) // Reading data...
-  var folderNamePath = makeFolder(path, file) // new folder to save splitted files
-  saveFile(folderNamePath, file, fileData, flag) // saving files
+function splitObject(path, file, flag = 1, callback, keys = []) { // split large files into single elements
```
[the remark this answers is not in the document]
Comment (atherdon) [reply generator/writeFile.js]: @uniyalprashant9 please try to crack this issue at this time.
i'm stuck with my own code, so cannot jump here right now. We'll wait while Vadim will wake up and he will try to help you.

**Comment by atherdon** [comment]
just briefly read your comments.
i think last problem you mentioned related to paths.
because when something related to files or includes/imports - paths is always pain in the ass. 
So if I was you, i'll try to split things as simple as they can possibly be and try to console.log each thing(variable, returned value from method, etc)
hope it will help

File: generator/writeFile.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (uniyalprashant9) [reply generator/writeFile.js]: I will fix all codeclimate issues once i fix all errors with my code.

**Comment by vadim9999** [comment]
1. Now all functions that using writeFile broke it. Other big tasks will find out which function using it.
2 I think it was a problem with readAllFiles. Adding whitespaces make it more readable but it requires additional code. Ok changes were done.

**Comment by atherdon** [comment]
so what are our next steps should be?

**Comment by vadim9999** [comment]
Sorry, I can't understand what was the problem with reading without whitespaces. Code
```
const path = "/home/vadim/work/javascript/sd/src/data/UserGrocery/"
console.log(utils.readAllFiles(path))
```
In logs
```
[ [ { user_id: '3d2055bb-6502-11e9-941e-f95703ddfd94',
      grocery_id: '3d202e90-6502-11e9-941e-f95703ddfd94' },
    { user_id: '3d2055bb-6502-11e9-941e-f95703ddfd94',
      grocery_id: '3d2055a0-6502-11e9-941e-f95703ddfd94' },
    { user_id: '3d2055bc-6502-11e9-941e-f95703ddfd94',
      grocery_id: '3d2055a0-6502-11e9-941e-f95703ddfd94' },
    { user_id: '3d2055bc-6502-11e9-941e-f95703ddfd94',
      grocery_id: '3d2055a1-6502-11e9-941e-f95703ddfd94' },
    { user_id: '3d2055bd-6502-11e9-941e-f95703ddfd94',
      grocery_id: '3d2055a1-6502-11e9-941e-f95703ddfd94' },
    { user_id: '3d2055bd-6502-11e9-941e-f95703ddfd94',
      grocery_id: '3d2055a2-6502-11e9-941e-f95703ddfd94' } ] ]
```

**Comment by vadim9999** [comment]
Steps:
1. After merging this PR needs test functions using writeFile. Because my method used async and without adding whitespaces.
2. Add description into Readme with log output.

**Comment by atherdon** [comment]
can you create a separated issue for that kind of work?
still not clear for me - can this pull be merged, or we should wait while
that will be done?

>

**Comment by vadim9999** [comment]
Yes we can merge it

**Comment by uniyalprashant9** [comment]
Please wait, I have to fix codclimate issues.

**Comment by vadim9999** [comment]
Ok

**Comment by uniyalprashant9** [comment]
Without delay
image

With delay
image

I think i should add a small delay in callback.

**Comment by uniyalprashant9** [comment]
@atherdon  @vadim9999 
Adding a small delay worked.

image

**Comment by atherdon** [comment]
so we can merge your latest changes?

>

**Comment by uniyalprashant9** [comment]
Please wait a few mins, I am adding new changes then we can merge :)

**Comment by atherdon** [comment]
ok, buzz us when you think it will be done.

**Comment by uniyalprashant9** [comment]
@atherdon @vadim9999 
I have added new changes ( [screenshot](https://github.com/GroceriStar/sd/pull/577#issuecomment-506963358) )  and also updated readme.md
We can now merge this PR.

**Comment by uniyalprashant9** [comment]
@atherdon Please give me a new task.
When will we work on ML?

**Pull request merged**